arangod/Cluster/SynchronizeShard.cpp

```diff
@@ -413,110 +414,93 @@ static inline bool isStopping() {
-                                               std::string const& clientId, uint64_t rlid,
-                                               bool soft, double timeout) {
+arangodb::Result SynchronizeShard::getReadLock(
+  std::string const& endpoint, std::string const& database,
+  std::string const& collection, std::string const& clientId,
+  uint64_t rlid, bool soft, double timeout) {
+
+  // This function can be implemented in a more robust manner for server
+  // versions > 3.4. Starting with 3.4 the POST requests to the read lock API
+  // terminates the server side thread as soon as the lock request comes in.
+  // The POST request thus is answered immediately back to the caller.
+  // The servers (<=3.3) with lower versions hold the POST request for as long
+  // as the corresponding DELETE_REQ has not been successfully submitted.
+  
+  using namespace std::chrono;
+  auto const start = steady_clock::now();
   auto cc = arangodb::ClusterComm::instance();
-  if (cc == nullptr) {  // nullptr only happens during controlled shutdown
-    return arangodb::Result(TRI_ERROR_SHUTTING_DOWN,
-                            "startReadLockOnLeader: Shutting down");
+
+  // nullptr only happens during controlled shutdown
+  if (cc == nullptr) {  
+    return arangodb::Result(
+      TRI_ERROR_SHUTTING_DOWN, "startReadLockOnLeader: Shutting down");
   }
 
   VPackBuilder body;
-  {
-    VPackObjectBuilder o(&body);
+  { VPackObjectBuilder o(&body);
     body.add(ID, VPackValue(std::to_string(rlid)));
     body.add(COLLECTION, VPackValue(collection));
     body.add(TTL, VPackValue(timeout));
-    body.add(StaticStrings::ReplicationSoftLockOnly, VPackValue(soft));
-  }
-
-  auto url = DB + database + REPL_HOLD_READ_LOCK;
+    body.add(REBOOT_ID, VPackValue(ServerState::instance()->getRebootId()));
+    body.add(StaticStrings::ReplicationSoftLockOnly, VPackValue(soft)); }
 
-  cc->asyncRequest(TRI_NewTickServer(), endpoint, rest::RequestType::POST, url,
-                   std::make_shared<std::string>(body.toJson()),
-                   std::unordered_map<std::string, std::string>(),
-                   std::make_shared<SynchronizeShardCallback>(this), timeout, true, timeout);
+  auto const url = DB + database + REPL_HOLD_READ_LOCK;
 
-  // Intentionally do not look at the outcome, even in case of an error
-  // we must make sure that the read lock on the leader is not active!
-  // This is done automatically below.
+  // Try to POST the lock body. If POST fails, we should just exit and retry
+  // SynchroShard anew. 
+  auto postres = cc->syncRequest(
+    TRI_NewTickServer(), endpoint, rest::RequestType::POST, url, body.toJson(),
+    std::unordered_map<std::string, std::string>(), timeout);
 
-  double sleepTime = 0.5;
-  size_t count = 0;
-  size_t maxTries = static_cast<size_t>(std::floor(600.0 / sleepTime));
-  while (++count < maxTries) {  // wait for some time until read lock established:
```

Comment:
Why don't we have to retry any more?
If I am not mistaken, if we do not get this lock for some reason we fall back to the cost-expensive collection comparison, so we shall try our best to get this lock.

Reply:
As far as I see this is called from `startReadLockOnLeader` and there is a retry loop around the call to that.

Reply:
yes. it is retried.